src/core/client_channel/retry_interceptor.cc

```diff
@@ -0,0 +1,391 @@
+                          }
+                          return status;
+                        }),
+                    [self]() { self->request_buffer_.Cancel(); });
+  });
+}
+
+void RetryInterceptor::Call::StartAttempt() {
+  if (current_attempt_ != nullptr) {
+    current_attempt_->Cancel();
+  }
+  auto current_attempt = call_handler_.arena()->MakeRefCounted<Attempt>(Ref());
+  current_attempt_ = current_attempt.get();
+  current_attempt->Start();
+}
+
+void RetryInterceptor::Call::MaybeCommit(size_t buffered) {
+  GRPC_TRACE_LOG(retry, INFO) << DebugTag() << " buffered:" << buffered << "/"
+                              << interceptor_->per_rpc_retry_buffer_size_;
+  if (buffered >= interceptor_->per_rpc_retry_buffer_size_) {
+    current_attempt_->Commit();
+  }
+}
+
+std::string RetryInterceptor::Call::DebugTag() {
+  return absl::StrFormat("%s call:%p", Activity::current()->DebugTag(), this);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// RetryInterceptor::Attempt
+
+RetryInterceptor::Attempt::Attempt(RefCountedPtr<Call> call)
+    : reader_(call->request_buffer()), call_(std::move(call)) {
+  GRPC_TRACE_LOG(retry, INFO) << DebugTag() << " retry attempt created";
+}
+
+RetryInterceptor::Attempt::~Attempt() { call_->RemoveAttempt(this); }
+
+auto RetryInterceptor::Attempt::ServerToClientGotInitialMetadata(
+    ServerMetadataHandle md) {
+  GRPC_TRACE_LOG(retry, INFO)
+      << DebugTag() << " get server initial metadata " << md->DebugString();
+  Commit();
+  call_->call_handler()->SpawnPushServerInitialMetadata(std::move(md));
+  return Seq(
+      ForEach(OutgoingMessages(&initiator_),
+              [call = call_](MessageHandle message) {
+                GRPC_TRACE_LOG(retry, INFO)
+                    << call->DebugTag() << " got server message "
+                    << message->DebugString();
+                return call->call_handler()->SpawnPushMessage(
+                    std::move(message));
+              }),
+      initiator_.PullServerTrailingMetadata(),
+      [call = call_](ServerMetadataHandle md) {
+        GRPC_TRACE_LOG(retry, INFO)
+            << call->DebugTag()
+            << " got server trailing metadata: " << md->DebugString();
+        call->call_handler()->SpawnPushServerTrailingMetadata(std::move(md));
+        return absl::OkStatus();
+      });
+}
+
+auto RetryInterceptor::Attempt::ServerToClientGotTrailersOnlyResponse() {
+  GRPC_TRACE_LOG(retry, INFO) << DebugTag() << " got trailers only response";
+  return Seq(
+      initiator_.PullServerTrailingMetadata(),
+      [self = Ref()](ServerMetadataHandle md) {
+        GRPC_TRACE_LOG(retry, INFO)
+            << self->DebugTag()
+            << " got server trailing metadata: " << md->DebugString();
+        auto pushback = self->call_->ShouldRetry(
+            *md,
+            [self = self.get()]() -> std::string { return self->DebugTag(); });
+        return If(
```

Comment:
ShouldRetry is not a promise, but the branches are.
`If` here creates a composite promise from the two branch promises... essentially:

```
bool branch_value_;
union {
  TrueBranchPromise true_branch_;
  FalseBranchPromise false_branch_;
};
```

so that each time the `If` promise is polled it can choose which branch promise was constructed and poll that.
This lets us create a branching construct without memory allocations that can remember which way it went all the way through the stack.

The other way of doing an `if` that certainly looks prettier:

```
Promise<T> Foo(bool k) {
  if (k) return []() { /* true branch */ }
  else return []() { /* false branch */ }
}
```

Note here that `Promise<T>` is really `AnyInvocable<Poll<T>>` and as such takes an allocation (in general) each time it's called. This is super appropriate for higher level control plane code (indeed I've got a PR queued for chaotic-good that does exactly that in the control plane to simplify the code), but for the data plane we prefer to avoid that.

Reply:
So here we're constructing a promise that will be polled later.

Based on a boolean, we want to return promise A or promise B.
These have different types.

There are two ways of implementing this:
Option 1: perform type erasure and have some type with `virtual Poll<T> DoPoll()=0`, and allocate memory to hold the promise that corresponds to the correct branch given the boolean - this implies allocation and virtual function calls every poll - great for control plan, not so much here
Option 2: capture the decision in a type. Ultimately we end up with `variant<A,B>` or something with the same semantics. The A or B-ness of the new type never changes, but it needs to know which way things went every poll so it can fetch out the right variant case.

(here we take Option 2 - if/when we do more control plane things we'll do some of Option 1 too - eg #38227 is forming part of a control plane for chaotic-good transports, and does extensive type erasure to keep things separated)

Reply:
Ah, okay -- so the goal here is really to avoid the separate allocation, not really to defer the evaluation of the conditional.

Thanks, that cleared things up for me!